src/android/java/io/jxcore/node/BtToServerSocket.java

```diff
@@ -53,61 +59,50 @@ public BtToServerSocket(BluetoothSocket socket, BtSocketDisconnectedCallBack han
         try {
-            mLocalHostAddress = (Inet4Address) Inet4Address.getByName("localhost");
+            Inet4Address mLocalHostAddress = (Inet4Address) Inet4Address.getByName("localhost");
             localHostSocket = new Socket(mLocalHostAddress, mHTTPPort);
 
-            if(localHostSocket != null){
-                mHTTPPort = localHostSocket.getPort();
-            }
-            print_debug("LocalHost addr: " + GetLocalHostAddress() + ", port: " + GetLocalHostPort());
+            mHTTPPort = localHostSocket.getPort();
+
+            print_debug("LocalHost address: " + GetLocalHostAddressAsString() + ", port: " + GetLocalHostPort());
 
-            print_debug("Get local streams");
             tmpInputStream = localHostSocket.getInputStream();
             tmpOutputStream = localHostSocket.getOutputStream();
 
         } catch (Exception e) {
             print_debug("Creating local input streams failed: " + e.toString());
             mStopped = true;
-            mHandler.Disconnected(that,"creating local input streams failed");
+            mHandler.Disconnected(that, "creating local input streams failed");
+            return;
         }
-        if(!mStopped) {
-            LocalInputStream = tmpInputStream;
-            LocalOutputStream = tmpOutputStream;
-
-            if (mmInStream != null && LocalInputStream != null
-                    && mmOutStream != null && LocalOutputStream != null
-                    && localHostSocket != null) {
-
-                SendingThread = new StreamCopyingThread(this, LocalInputStream, mmOutStream);
-                SendingThread.setDebugTag("--Sending");
-                SendingThread.start();
-
-                ReceivingThread = new StreamCopyingThread(this, mmInStream, LocalOutputStream);
-                ReceivingThread.setDebugTag("--Receiving");
-                ReceivingThread.start();
-                print_debug("Stream Threads are running");
-            } else {
-                mStopped = true;
-                print_debug("at least one stream is null");
-                mHandler.Disconnected(that,"at least one stream is null");
-            }
+
+        LocalInputStream = tmpInputStream;
+        LocalOutputStream = tmpOutputStream;
+
+        if (mmInStream == null || LocalInputStream == null || mmOutStream == null || LocalOutputStream == null || localHostSocket == null) {
+            mStopped = true;
+            print_debug("at least one stream is null");
+            mHandler.Disconnected(that, "at least one stream is null");
+            return;
         }
 
+        SendingThread = new StreamCopyingThread(this, LocalInputStream, mmOutStream);
+        SendingThread.setDebugTag("--Sending");
+        SendingThread.start();
+
+        ReceivingThread = new StreamCopyingThread(this, mmInStream, LocalOutputStream);
+        ReceivingThread.setDebugTag("--Receiving");
+        ReceivingThread.start();
+
         print_debug("--DoOneRunRound ended");
     }
 
```

Comment:
StreamCopyError in this file is literally line for line identical can you please factor it out into some common class? We don't want to repeat code. It just invites bugs. In fact large parts of BtToRequestSocket and BtToServerSocket are identical include StreamCopyError, big chunks of run(), GetLocalHostPort, GetLocalHostAddressAsString, Stop(), SetIdAddressAndName, GetPeerId, GetPeerName, GetPeerAddress and print_Debug. Can you please factor all these methods out into a common base class and inherit from there?